// @adobe/exc-app/userprofile.ts
/*************************************************************************
 * Copyright 2020 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 **************************************************************************/
/**
 * APIs that let solutions interact with User Profile menu.
 *
 * ***Import:***
 *
 * ```typescript
 * import userProfile from '@adobe/exc-app/userProfile';
 * ```
 *
 * ***Default export:***
 *
 * [UserProfileApi](../interfaces/userprofile.userprofileapi.md#interface-userprofileapi)
 *
 * ***Usage:***
 *
 * ```typescript
 * import userProfile from '@adobe/exc-app/userProfile';
 *
 * userProfile.setButtons([{
 *   callback: () => {},
 *   id: 'button',
 *   label: 'Button'
 * }]);
 * ```
 *
 * This configuration is a method only for solutions that need to manage alternative User Profile flows not covered by the Shell.
 * @packageDocumentation
 * @module userProfile
 */
import {connect} from './src/Global';

export interface Button {
  /**
   * Callback function to be enacted onclick.
   */
  callback: (value?: any) => void;
  /**
   * Text shown on the button.
   */
  label: string;
  /**
   * Unique identifier for the button.
   */
  id: string;
}

/**
 * A maximum of 3 buttons can be supplied.
 */
export type ButtonArray = [] | [Button] | [Button, Button] | [Button, Button, Button];

/**
 * Defines the user profile api.
 */
export interface UserProfileApi {
  /**
   * Adds custom button(s) to the profile dropdown and enacts a callback onclick.
   * Custom user profile buttons may be reset by entering a null parameter.
   * A maximum of 3 custom buttons will be displayed in the User Profile popover.
   *
   * ```typescript
   * userProfile.setButtons([
   *   {
   *     callback: () => {},
   *     label: 'Alpha',
   *     id: 'alpha'
   *   },
   *   {
   *     callback: () => {},
   *     label: 'View My Profile',
   *     id: 'viewprofile'
   *   }
   * ]);
   * ```
   */
  setButtons(buttons: ButtonArray | null): void;
}

const userProfile = connect('userProfile', [
  ['setButtons', true]
]);

export default userProfile;
