// @adobe/exc-app/internal.js
"use strict";
/*************************************************************************
 * Copyright 2020 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 **************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Internal = void 0;
const Global_1 = require("./src/Global");
/**
 * @ignore
 */
class Internal {
    static configureMetrics(config) {
        return Global_1.getImpl('internal').configureMetrics(config);
    }
    static configureNetwork(config) {
        return Global_1.getImpl('internal').configureNetwork(config);
    }
    static flush() {
        return Global_1.getImpl('internal').flush();
    }
    static setApplication(app) {
        return Global_1.getImpl('internal').setApplication(app);
    }
    static setFilter(filterFunction) {
        return Global_1.getImpl('internal').setFilter(filterFunction);
    }
    static setUser(user, version) {
        return Global_1.getImpl('internal').setUser(user, version);
    }
    static clearUser(version) {
        return Global_1.getImpl('internal').clearUser(version);
    }
}
exports.Internal = Internal;
