// @adobe/exc-app/nps.ts
/*************************************************************************
 * Copyright 2020 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 **************************************************************************/
/**
 * APIs that let solutions interact with the NPS (Net Promoter Score) widget component, the interface for recording user experience within a specific solution.
 *
 * ***Import:***
 *
 * ```typescript
 * import nps from '@adobe/exc-app/nps';
 * ```
 *
 * ***Default export:***
 *
 * [NpsApi](../interfaces/nps.npsapi.md#interface-npsapi)
 *
 * ***Usage:***
 *
 * ```typescript
 * import nps from '@adobe/exc-app/nps';
 *
 * nps.config = {
 *   enabled: true,
 *   sampling: 20,
 *   showToAdobeUsers: false
 * };
 * ```
 * @packageDocumentation
 * @module nps
 */
import {connect} from './src/Global';

export interface NpsApiProperties {
  /**
   * An object of nps configuration attributes for solution.
   * By default the NPS widget is disabled. In order to enable NPS for a solution, the enabled key in the solution configuration should be set to true.
   * NPS configuration also contains a sampling key which describes the percentage of users that will see the NPS widget.
   * Whether NPS will be shown to Adobe users depends of the value of 'showToAdobeUsers' key. It is set to true by default.
   */
  config: {
    /**
     * Flag that enables nps for solution.
     */
    readonly enabled?: boolean;

    /**
     * Number that shows percent of user logins that can show nps.
     */
    readonly sampling?: number;

    /**
     * Flag that enables showing nps survey to adobe users.
     */
    readonly showToAdobeUsers?: boolean;
  };
}

/**
 * Defines Nps widget APIs available to solutions.
 */
export interface Nps extends NpsApiProperties {
  /**
   * Triggers calculations to evaluate whether to display NPS survey to the user
   * and then render the NPS widget if so.
   *
   * ```typescript
   * nps.triggerNps();
   * ```
   */
  triggerNps(): void;
}
const nps = connect('nps', [
  ['config'],
  ['triggerNps', true]
]);
export default nps;
