// @adobe/exc-app/metrics/User.ts
/*************************************************************************
 * Copyright 2020 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 **************************************************************************/
import Metric from './Metric';

/**
 * @ignore
 */
export default interface User extends Metric {
  accountType?: string; // Analytics Data Layer: The account level of the user for your given product.
  attributes?:  any; // Analytics Data Layer: additional values stored in key:value pairs with the value ALWAYS being strings.
  authOrigin?:  string; // Analytics Data Layer: The origin of where the user logged in or "anonymous" if the user is not logged in.
  authSystem?:  string; // Analytics Data Layer: The system used to provide the user.id and user.corpId.
  displayName?: string; // firstname lastname - for convenience in viewing/searching
  groupId?:     string; // user's group identifier, e.g. currently selected IMS org, or a solution user group
  groupName?:   string; // current effective group for the user, e.g. current IMS org, for viewing/searching
  ipAddress?:   string; // as viewed by the metrics gateway - an indicator of the user's location
  language?:    string; // Analytics Data Layer: The language/locale setting the user has currently selected within the UI.
  privileges?:  string[]; // Analytics Data Layer: Privileges that the user currently has access to as relating to the product as a whole.
  sessionId?:   string; // user's session identifier, e.g. IMS session or browser session header/cookie
  userAgent?:   string; // a client identifier, e.g. the User-Agent request header
  userId?:      string; // user's login string, e.g. AdobeId or username, as submitted to the auth system
}
